Add dynamic type support documentation and considerations (feature PR) (#954)
hashicorp/terraform-plugin-framework: website/docs/plugin/framework/handling-data/dynamic-data.mdx

Author: This page I think needs the most review/scrutiny. I wasn't sure the exact terminology or how much detail to provide, so please feel free to make suggestions around structure as well.

My initial approach was to word all the considerations starting with the negative impact followed by solution

Author: I was not sure if we should include some example scenarios on when you **should** use a dynamic type? Does anyone have a list of those exact scenarios that couldn't be covered by one or more static types?

---
page_title: 'Plugin Development - Framework: Handling Data - Dynamic Data'
description: >-
  How to handle data when utilizing dynamic attributes or dynamic function parameters.
---

# Dynamic Data

<Note>

> Dynamic data handling information is only relevant when utilizing a [dynamic attribute](/terraform/plugin/framework/handling-data/attributes/dynamic), [dynamic parameter](/terraform/plugin/framework/functions/parameters/dynamic) or [dynamic return](/terraform/plugin/framework/functions/returns/dynamic).

Author: Good point, I've removed this note and reworked the first paragraph to list every possible scenario you can utilize dynamics in 29c79cc ⚡

Static types should always be preferred over dynamic types, when possible.

</Note>

Dynamic data handling uses the [framework dynamic type](/terraform/plugin/framework/handling-data/types/dynamic) to communicate to Terraform that the value type of an [attribute](/terraform/plugin/framework/handling-data/attributes/dynamic), [function parameter](/terraform/plugin/framework/functions/parameters/dynamic), or [function return](/terraform/plugin/framework/functions/returns/dynamic), will be determined at runtime. This allows a provider developer to handle multiple value types of data with a single attribute/parameter/return.

Using dynamic data has a negative impact on practitioner experience when using Terraform and downstream tooling, like practitioner configuration editor integrations. Dynamics do not change how [Terraform's static type system](/terraform/language/expressions/types) behaves and all data consistency rules are applied the same as static types. Provider developers should understand all the below [considerations](#considerations) when creating a provider with a dynamic type.

> It's recommended that provider developers only use dynamics when using one or more static types is not a suitable alternative.

Reviewer: Nit: This is essentially reiterating the note callout about. Here are stylistic suggestions to follow the HashiCorp writing guidance (e.g. no contractions), but maybe this sentence should be moved/removed and only in the callout.

```suggestion
Only use a dynamic type when there is not a suitable static type alternative.
```

```go
func (r ThingResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
    resp.Schema = schema.Schema{
        Attributes: map[string]schema.Attribute{
            "example_attribute": schema.DynamicAttribute{
                Required: true,
                // ... potentially other fields ...
            },
            // ... potentially other attributes ...
        },
    }
}
```

An example of handling every possible Terraform type that could be provided to a configuration would be:

```go
    // Example data model definition
    // type ExampleModel struct {
    //   ExampleAttribute types.Dynamic `tfsdk:"example_attribute"`
    // }
    switch value := data.ExampleAttribute.UnderlyingValue().(type) {
    case types.Bool:
        // Handle boolean value
    case types.Number:
        // Handle number value
    case types.List:
        // Handle list value
    case types.Map:
        // Handle map value
    case types.Object:
        // Handle object value
    case types.Set:
        // Handle set value
    case types.String:
        // Handle string value
    case types.Tuple:
        // Handle tuple value
    }
```

> When writing test configurations and debugging provider issues, developers will also want to understand how Terraform represents [complex type literals](/terraform/language/expressions/type-constraints#complex-type-literals). For example, Terraform does not provide any way to directly represent lists, maps, or sets.

Reviewer 2: The linked page provides an excellent level of detail, I'm just wondering whether it might be worth saying something about the relationship between framework types, and the Terraform representation of complex type literals, as the former will likely be familiar to provider developers, but the latter may not be.

        },
    }
}
```

If a practitioner configures this resource as:

```hcl
resource "examplecloud_thing" "example" {
  # This literal expression is a tuple[string, string]
  example_attribute = ["one", "two"]
}
```

Then the exact type must be planned and stored in state during `apply` as a [tuple](/terraform/plugin/framework/handling-data/types/tuple) with two [string](/terraform/plugin/framework/handling-data/types/string) element types. If provider code attempts to store this attribute as a different type, like a [list](/terraform/plugin/framework/handling-data/types/list) of strings, even with the same data values, Terraform will produce an error during apply:

```bash
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to examplecloud_thing.example, provider "provider[\"TYPE\"]" produced an unexpected new value: .example_attribute: wrong final value type: tuple required.
│ 
│ This is a bug in the provider, which should be reported in the providers own issue tracker.
```

If a practitioner configures this same resource as:

```hcl
resource "examplecloud_thing" "example" {
  example_attribute = tolist(["one", "two"])
}
```

> Then the exact type must be planned and stored in state during `apply` as a [list](/terraform/plugin/framework/handling-data/types/list) of strings.

Reviewer 2: Perhaps expand this section to say a little more, or provide examples, about the handling/behaviour of Terraform list/tuple, map/object, and framework list, map, object and set types in this regard?

In this example, a resource schema defines a top level required [dynamic attribute](/terraform/plugin/framework/handling-data/attributes/dynamic) named `example_attribute`:

```go
func (r ThingResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
    resp.Schema = schema.Schema{
        Attributes: map[string]schema.Attribute{
            "example_attribute": schema.DynamicAttribute{
                Required: true,
                // ... potentially other fields ...
            },
            // ... potentially other attributes ...
        },
    }
}
```

An example of handling every possible Terraform type that could be provided to a configuration would be:

```go
    // Example data model definition
    // type ExampleModel struct {
    //   ExampleAttribute types.Dynamic `tfsdk:"example_attribute"`
    // }
    switch value := data.ExampleAttribute.UnderlyingValue().(type) {
    case types.Bool:
        // Handle boolean value
    case types.Number:

>         // Handle number value

Reviewer: Oh! There is potential for a good shout here or somewhere that framework-specific number types like `Float64` and `Int64` will _not_ ever be in a dynamic value since Terraform's type system _only_ has the single number type.

Author: Added in de4201b ⚡
